src/counts.jl

```diff
@@ -231,16 +231,83 @@ end
 
 
 """
-    addcounts!(dict, x[, wv])
+    addcounts!(dict, x[, wv]; alg = :auto)
 
 Add counts based on `x` to a count map. New entries will be added if new values come up.
 If a weighting vector `wv` is specified, the sum of the weights is used rather than the
 raw counts.
+
+`alg` can be one of:
+- `:auto` (default): if `StatsBase.radixsort_safe(eltype(x)) == true` then use
+                     `:radixsort`, otherwise use `:dict`.
+
+- `:radixsort`:      if `radixsort_safe(eltype(x)) == true` then use the
+                     [radix sort](https://en.wikipedia.org/wiki/Radix_sort)
+                     algorithm to sort the input vector which will generally lead to
+                     shorter running time. However the radix sort algorithm creates a
+                     copy of the input vector and hence uses more RAM. Choose `:dict`
+                     if the amount of available RAM is a limitation.
+
+- `:dict`:           use `Dict`-based method which is generally slower but uses less
+                     RAM and is safe for any data type.
 """
-function addcounts!(cm::Dict{T}, x::AbstractArray{T}) where T
+function addcounts!(cm::Dict{T}, x::AbstractArray{T}; alg = :auto) where T
+    # if it's safe to be sorted using radixsort then it should be faster
+    # albeit using more RAM
+    if radixsort_safe(T) && (alg == :auto || alg == :radixsort)
+        addcounts_radixsort!(cm, x)
+    elseif alg == :radixsort
+        throw(ArgumentError("`alg = :radixsort` is chosen but type `radixsort_safe($T)` did not return `true`; use `alg = :auto` or `alg = :dict` instead"))
+    else
+        addcounts_dict!(cm,x)
+    end
+    return cm
+end
+
+"""Dict-based addcounts method"""
+function addcounts_dict!(cm::Dict{T}, x::AbstractArray{T}) where T
     for v in x
-        cm[v] = get(cm, v, 0) + 1
+        index = Base.ht_keyindex2(cm, v)
```

Comment:
Looks like `ht_keyindex2` no longer exists in 0.7; there's `ht_keyindex` and `ht_keyindex2!`. I'm guessing the latter is what's appropriate here?